pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,419 @@
+	var ret []types.UID
+	for _, ref := range refs {
+		ret = append(ret, ref.UID)
+	}
+	return ret
+}
+
+func (gc *GarbageCollector) attempToDeleteItem(item *node) error {
+	glog.V(2).Infof("processing item %s", item.identity)
+	// "being deleted" is an one-way trip to the final deletion. We'll just wait for the final deletion, and then process the object's dependents.
+	if item.isBeingDeleted() && !item.isDeletingDependents() {
+		glog.V(5).Infof("processing item %s returned at once", item.identity)
+		return nil
+	}
+	// TODO: It's only necessary to talk to the API server if this is a
+	// "virtual" node. The local graph could lag behind the real status, but in
+	// practice, the difference is small.
+	latest, err := gc.getObject(item.identity)
+	if err != nil && errors.IsNotFound(err) {
+		// the GraphBuilder can add "virtual" node for an owner that doesn't
+		// exist yet, so we need to enqueue a virtual Delete event to remove
+		// the virtual node from GraphBuilder.uidToNode.
+		glog.V(5).Infof("item %v not found, generating a virtual delete event", item.identity)
+		gc.generateVirtualDeleteEvent(item.identity)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	if latest.GetUID() != item.identity.UID {
+		glog.V(5).Infof("UID doesn't match, item %v not found, generating a virtual delete event", item.identity)
+		gc.generateVirtualDeleteEvent(item.identity)
+		return nil
+	}
+
+	// TODO: attemptToOrphanWorker() routine is similar. Consider merging
+	// attemptToOrphanWorker() into attempToDeleteItem() as well.
+	if item.isDeletingDependents() {
+		return gc.processDeletingDependentsItem(item)
+	}
+
+	// compute if we should delete the item
+	ownerReferences := latest.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		glog.V(2).Infof("object %s's doesn't have an owner, continue on next item", item.identity)
+		return nil
+	}
+
+	solid, dangling, waitingForDependentsDeletion, err := gc.classifyReferences(item, ownerReferences)
+	if err != nil {
+		return err
+	}
+	glog.V(5).Infof("classify references of %s.\nsolid: %#v\ndangling: %#v\nwaitingForDependentsDeletion: %#v\n", item.identity, solid, dangling, waitingForDependentsDeletion)
+
+	switch {
+	case len(solid) != 0:
+		glog.V(2).Infof("object %s has at least one existing owner, will not garbage collect", item.identity)
+		if len(dangling) != 0 || len(waitingForDependentsDeletion) != 0 {
+			glog.V(2).Infof("remove dangling references %#v and waiting references %#v for object %s", dangling, waitingForDependentsDeletion, item.identity)
+		}
+		patch := deleteOwnerRefPatch(item.identity.UID, append(ownerRefsToUIDs(dangling), ownerRefsToUIDs(waitingForDependentsDeletion)...)...)
+		_, err = gc.patchObject(item.identity, patch)
+		return err
+	case len(waitingForDependentsDeletion) != 0 && len(item.dependents) != 0:
+		for dep := range item.dependents {
+			if dep.isDeletingDependents() {
+				// this circle detection has false positives, we need to
```

Comment:
Not sure what an "indirect cycle" is, could you give an example? This code detects a cycle when processing the last link that completes a cycle.

An unfortunate graph topology + an unfortunate processing order could result a false negative. I think it will be rare in practice. We can switch to a full-fledged cycle detection if that becomes a problem.